### ISSUE TITLE
Custom language: manual attributes in gazelle rules

### ISSUE THREAD

**Issue by tgolsson** [bazelbuild/bazel-gazelle#1230]
### What version of gazelle are you using?

v0.23.1-0.20210528183849-d038863ba2e0 is the exact code version in use.

### What did you do?

I'm implementing a custom language plugin to automatically generate rules for another rules package that doesn't have such. This rules package doesn't support Gazelle, so I'm simply trying a best-effort solution to generate 90% of the data.

In particular, since some deps don't get indexed by the RuleIndex and some data can't be scanned from files I need to have users manually adding a few dependencies. The way I've set this up is such that I generate srcs + deps, and then users add an `extras` attribute. However, this `extras` attribute is *sometimes* stripped from the generated rules, and sometimes kept. 

I believe this is related to rules-merging, but I'm not sure why it happens. So for example, I'll have something like
```python
xyz_rule(
    name = "bundle",
    deps = [ ... ],
    srcs = [ ... ],
    extras = [... ],
)
```

and upon regeneration I'll be left with

```python
xyz_rule(
    name = "bundle",
    deps = [ ... ],
    srcs = [ ... ],
)
```

which loses data permanently. I haven't isolated a trigger, but if it happens, I can repeatedly revert the file and reproduce it. Changing the `language.Kinds()` entry for `xyz_rule` will affect it, but it seems like it's mostly about *changing* it and not that much about what the `Kinds()` contains (i.e., it feels like there's some state here that I'm not seeing that affects it).

The desired solution is to have some attributes (`extras`) be authoritative as they are *in the file* and some fields be authoritative as they are when generated (`srcs`, `deps`). Is this possible?

**Comment by achew22** [comment]
Hey @tgolsson, sorry to be slow on the response here. Is there any chance you can put the code online somewhere? This is one of those things where it's easy to start tilting at windmills and a concrete example would really help it out.

It completely sounds like an issue with `Kinds()`, but without some visibility into it, it's going to be hard.

Sorry to not e more helpful

**Comment by tgolsson** [comment]
@achew22 No worries, open source is always slow. Happy to get any response at all :D I'll try to clean it up the full thing between meetings tomorrow and add here. In the meantime; this is the Kinds rule:

```go
func (*webpackBundleLang) Kinds() map[string]rule.KindInfo {
	return map[string]rule.KindInfo{
		"webpack_bundle": {
			NonEmptyAttrs:  map[string]bool{"deps": true, "srcs": true},
			MergeableAttrs: map[string]bool{"srcs": true},
		},
                ...
	}
}
```

and my goal is to have a rule like 

```starlark
webpack_bundle(
    # auto
    name = "foo",
    deps = [ ... ],
    srcs = [ ... ],
    # manual
    extras = [ ... ],
)
```

And currently the `extras` sometimes disappears -- I haven't spent time on this since I posted the issue; but it looks like it's *mostly* kept but then sometimes the full file will be rewritten (i.e., all rules and fields are sorted/moved) and the manual fields get forgotten.

**Comment by achew22** [comment]
Are there any inputs for which it deterministically either makes it work or completely removes them?

**Comment by tgolsson** [comment]
@achew22 No; I'd not say so. The code isn't random; given a specific repository state it'll consistently leave it or not leave it. I just can't tell you which state *specifically* that triggers it.

**Comment by achew22** [comment]
We recently added a nice testing framework where you can set up inputs in `BUILD.in` and test against `BUILD.out` files that's exposed to the outside world using the [gazelle_generation_test](https://github.com/bazelbuild/bazel-gazelle/blob/master/extend.md#gazelle_generation_test) macro. We have some usage examples [here](https://github.com/bazelbuild/bazel-gazelle/tree/master/tests). It might be worth sitting down with a copy of `ibazel` (or `watch bazel test :target`) and hammering out a test case that exhibits this behavior so we can have a more precise conversation about it.
